ddl/ddl_api.go

```diff
@@ -2809,6 +2809,156 @@ func (d *ddl) DropTablePartition(ctx sessionctx.Context, ident ast.Ident, spec *
 	return errors.Trace(err)
 }
 
+func checkFieldTypeCompatible(ft *types.FieldType, other *types.FieldType) bool {
+	// int(1) could match the type with int(8)
+	partialEqual := ft.Tp == other.Tp &&
+		ft.Decimal == other.Decimal &&
+		ft.Charset == other.Charset &&
+		ft.Collate == other.Collate &&
+		(ft.Flen == other.Flen || ft.StorageLength() != types.VarStorageLen) &&
+		mysql.HasUnsignedFlag(ft.Flag) == mysql.HasUnsignedFlag(other.Flag) &&
```

Comment:
Sorry, why not use `ft.Flag == other.Flag` to compare directly.
BTW, below SQL will success in This PR but failed in MySQL5.7

```sql
drop table if exists t1,t;
create table t (a int not null, b int, index (a)) partition by hash(a) partitions 3;
insert into t values (1,1),(2,2),(3,3),(4,4);
create table t1 (a int key, b int, index (a));
alter table t exchange partition p0 with table t1;
```

Reply:
It seems that `(a int key)` equals to `(a int not null)` in `tidb`. To compatible with mysql, there are the other attribution that we needn't to compare with such as `default value`.

Reply:
```sql
mysql> create table t1 (a int key, b int, index (a));
Query OK, 0 rows affected (0.06 sec)

mysql> show create table t1;
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                             |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (
  `a` int(11) NOT NULL,
  `b` int(11) DEFAULT NULL,
  PRIMARY KEY (`a`),
  KEY `a` (`a`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)

```

Reply:
Still need to compare more flags such as `PriKeyFlag`